src/main/java/org/opentripplanner/graph_builder/module/ned/ElevationModule.java

```diff
@@ -80,55 +129,175 @@ public ElevationModule(ElevationGridCoverageFactory factory, double elevationUni
+                // first try to find a cached value if possible
+                if (cachedElevations != null) {
+                    PackedCoordinateSequence coordinateSequence = cachedElevations.get(
+                        PolylineEncoder.createEncodings(edgeGeometry).getPoints()
+                    );
+                    if (coordinateSequence != null) {
+                        // found a cached value! Set the elevation profile with the pre-calculated data.
+                        setEdgeElevationProfile(swee, coordinateSequence, graph);
+                        edgeNeedsProcessing = false;
+                    }
+                }
+            }
+
+            if (edgeNeedsProcessing) {
+                // Needs full calculation. Calculate with a thread-specific coverage to avoid waiting for any locks on
+                // coverage instances in other threads.
+                processEdge(swee, edgeGeometry, getThreadSpecificCoverageInstance());
+            }
+            int curNumProcessed = nEdgesProcessed.addAndGet(1);
+            if (curNumProcessed % 50000 == 0) {
+                log.info("set elevation on {}/{} edges", curNumProcessed, totalElevationEdges);
+            }
+        }
+
+        /**
+         * Get the thread-specific Coverage instance to avoid multiple threads waiting for a lock to be released on a
+         * Interpolator2D instance.
+         */
+        private Coverage getThreadSpecificCoverageInstance () {
+            long currentThreadId = Thread.currentThread().getId();
+            Coverage threadSpecificCoverage = coveragesForThread.get(currentThreadId);
+            if (threadSpecificCoverage == null) {
+                // Synchronize the creation of the Thread-specific Coverage instances to avoid potential locks that
+                // could arise from downstream classes that have synchronized methods.
+                synchronized (coveragesForThread) {
+                    // Get a new Coverage instance from the module's ElevationGridCoverageFactory.
+                    threadSpecificCoverage = gridCoverageFactory.getGridCoverage();
+                    // The Coverage instance relies on some synchronized static methods shared across all threads that
+                    // can cause deadlocks if not fully initialized. Therefore, make a single request for the first
+                    // point on the edge to initialize these other items.
+                    Coordinate firstEdgeCoord =  swee.getGeometry().getCoordinates()[0];
+                    double[] dummy = new double[1];
+                    threadSpecificCoverage.evaluate(
+                        new DirectPosition2D(GeometryUtils.WGS84_XY, firstEdgeCoord.x, firstEdgeCoord.y),
+                        dummy
+                    );
+                }
+                coveragesForThread.put(currentThreadId, threadSpecificCoverage);
+            }
+            return threadSpecificCoverage;
+        }
```

Comment:
It looks like the `coveragesForThread` is a map for each Thread to store some "data", another way to achieve it would be to create a `ForkJoinWorkerThreadFactory` which return a special thread with this data set. Then the map is absolute and the data only visible to the "correct" thread.